components/eamxx/src/diagnostics/zonal_avg.cpp

```diff
@@ -0,0 +1,190 @@
+              // TODO: check if tenary is ok here (if not, multiply by flag)
+              int flag = (lat_lower <= lat_view(i)) && (lat_view(i) < lat_upper);
+              val += flag ? weight_view(i) * field_view(i) : sp(0.0);
+            }, result_view(lat_i));
+        });
+    } break;
+    case 2: {
+      const int d1 = result_layout.dim(1);
+      auto field_view = field.get_view<const Real **>();
+      auto result_view = result.get_view<Real **>();
+      TeamPolicy team_policy =
+        ESU::get_default_team_policy(lat_num * d1, ncols);
+      Kokkos::parallel_for("compute_zonal_sum_" + field.name(), team_policy,
+        KOKKOS_LAMBDA(const TeamMember& tm) {
+          const int idx = tm.league_rank();
+          const int d1_i = idx / lat_num;
+          const int lat_i = idx % lat_num;
+          const Real lat_lower = sp(-90.0) + lat_i * lat_delta;
+          const Real lat_upper = lat_lower + lat_delta;
+          Kokkos::parallel_reduce(Kokkos::TeamVectorRange(tm, ncols),
+            KOKKOS_LAMBDA(int i, Real& val) {
+              int flag = (lat_lower <= lat_view(i)) && (lat_view(i) < lat_upper);
+              // TODO: check if tenary is ok here (if not, multiply by flag)
+              val += flag ? weight_view(i) * field_view(i,d1_i) : sp(0.0);
+            }, result_view(lat_i,d1_i));
+        });
+    } break;
+    case 3: {
+      const int d1 = result_layout.dim(1);
+      const int d2 = result_layout.dim(2);
+      auto field_view = field.get_view<const Real ***>();
+      auto result_view = result.get_view<Real ***>();
+      TeamPolicy team_policy =
+        ESU::get_default_team_policy(lat_num * d1 * d2, ncols);
+      Kokkos::parallel_for("compute_zonal_sum_" + field.name(), team_policy,
+        KOKKOS_LAMBDA(const TeamMember& tm) {
+          const int idx = tm.league_rank();
+          const int d1_i = idx / (lat_num * d2);
+          const int idx2 = idx % (lat_num * d2);
+          const int d2_i = idx2 / lat_num;
+          const int lat_i = idx2 % lat_num;
+          const Real lat_lower = sp(-90.0) + lat_i * lat_delta;
+          const Real lat_upper = lat_lower + lat_delta;
+          Kokkos::parallel_reduce(Kokkos::TeamVectorRange(tm, ncols),
+            KOKKOS_LAMBDA(int i, Real& val) {
+              int flag = (lat_lower <= lat_view(i)) && (lat_view(i) < lat_upper);
+              // TODO: check if tenary is ok here (if not, multiply by flag)
+              val += flag ? weight_view(i) * field_view(i,d1_i,d2_i) : sp(0.0);
+            }, result_view(lat_i,d1_i,d2_i));
+        });
+    } break;
+    default:
+      EKAT_ERROR_MSG("Error! Unsupported field rank for zonal averages.\n");
+  }
+
+  if (comm) {
+    // TODO: use device-side MPI calls
+    // TODO: the dev ptr causes problems; revisit this later
+    // TODO: doing cuda-aware MPI allreduce would be ~10% faster
+    Kokkos::fence();
+    result.sync_to_host();
+    comm->all_reduce(result.template get_internal_view_data<Real, Host>(),
```

Comment:
Just so we're clear, so far this diagnostic will NOT be PE-invariant. That said, @oksanaguba is working on implementing bridges to the repro_sum f90 module, so we can revisit this later. The horiz_avg diag already suffers from the same issue.

Reply:
That makes sense: the horiz_avg diag is the one I modeled this diag after.